### DISCUSSION TITLE
How to avoid installing the same `node_modules` folder in every extension

### DISCUSSION THREAD

**Discussion by Pachat** [directus/directus#11735] [Archive]
Scaffolding an extension of any type with
`npm init directus-extension`
installs the same `node_modules` folder of ~115 Mb. for each extension you create.

Is there a way to reuse one unique `node_modules` folder for all the extensions of a Directus project?

**Comment by joselcvarela** [comment]
That issue is not related with Directus itself but with npm.
Anyway you can achieve that by using npm workspaces.

**Comment by Pachat** [comment]
Well in fact, I ended without using workspaces, locally on Windows using the CLI:

In a root directory `directusroot` for all my directus applications, install **globally** directus
```
npm init --yes
npm i -g -D directus

```
Then, in `directusroot` run a project installation 
`npm init directus-project myproject` 
Then **remove** the `directusroot\myproject\node_modules` directory that directus-project installed
And run the app with `http://localhost:8055`. It works!

To use the **sdk** to create an extension, first install it **globally** in the root directory `directusroot`
`npm i -g -D @directus/extension-sdk`
and, in `directusroot\myproject\extensions\displays` for instance,  to create `mydisplay`, run the extension scaffolding installation
`npm init directus-extension`
Then **remove** the `directusroot\myproject\extensions\displays\mydisplay\node_modules` directory that directus-project installed
And build with `npm run build`. It works!
Finally copy `directusroot\myproject\extensions\displays\mydisplay\dist\index.js` to 
`myproject\extensions\displays\mydisplay\index.js`
and in `directusroot\myproject`, stop directus and restart. It works!

- The benefit installing globally is there is only one `node_modules` at the root level, shared by every Directus project and every extension.
- In such a situation, it does not look necessary to have a `node_modules` inside every Directus project and inside every extension.
- _This method to share only one `node_modules` is of course useful only if all projects run with the same directus version._

--

At the end, the installation follows this tree:

directusroot
∟ app1
&nbsp; &nbsp;∟ ~~node_modules~~
&nbsp; &nbsp;∟ extensions
&nbsp; &nbsp;&nbsp; &nbsp;∟ displmays
&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;∟ mydisplay
&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;∟ ~~node_modules~~
∟ app2
∟ node_modules

In case it helps, here is my `directusroot\package.json`
```
{
	"name": "Directus",
	"version": "1.0.0",
	"description": "Directus development",
	"main": "index.js",
	"scripts": {
		"build": "directus-extension build",
		"test": "echo \"Error: no test specified\" && exit 1"
	},
	"directus:extension": {
		"type": "module",
		"path": "dist/index.js",
		"source": "src/index.js",
		"host": "^9.5.2"
	},
	"keywords": [],
	"author": "",
	"license": "ISC",
	"dependencies": {
		"directus": "^9.5.2",
		"@directus/extensions-sdk": "9.5.2",
		"sqlite3": "^5.0.2",
		"vue": "^3.2.31"
	}
}

```
